core/src/test/java/de/danoeh/antennapod/core/storage/ItemEnqueuePositionCalculatorTest.java

```diff
@@ -0,0 +1,279 @@
+            // (rather than the expected positions)
+            return Arrays.asList(new Object[][] {
+                    {"download order test, enqueue default",
+                            concat(QUEUE_DEFAULT_IDS, 101L),
+                            concat(QUEUE_DEFAULT_IDS, list(101L, 102L)),
+                            BACK, QUEUE_DEFAULT, ID_CURRENTLY_PLAYING_NULL},
+                    {"download order test, enqueue at front (currently playing has no effect)",
+                            concat(101L, QUEUE_DEFAULT_IDS),
+                            concat(list(101L, 102L), QUEUE_DEFAULT_IDS),
+                            FRONT, QUEUE_DEFAULT, 11L}, // 11 is at the front, currently playing
+                    {"download order test, enqueue after currently playing",
+                            list(11L, 101L, 12L, 13L, 14L),
+                            list(11L, 101L, 102L, 12L, 13L, 14L),
+                            AFTER_CURRENTLY_PLAYING, QUEUE_DEFAULT, 11L}  // 11 is at the front, currently playing
+            });
+        }
+
+        @Parameter
+        public String message;
+
+        @Parameter(1)
+        public List<Long> idsExpectedAfter101;
+
+        @Parameter(2)
+        public List<Long> idsExpectedAfter102;
+
+        @Parameter(3)
+        public EnqueueLocation options;
+
+        @Parameter(4)
+        public List<FeedItem> queueInitial;
+
+        @Parameter(5)
+        public long idCurrentlyPlaying;
+
+        @Test
+        public void testQueueOrderWhenDownloading2Items() {
+
+            // Setup class under test
+            //
+            ItemEnqueuePositionCalculator calculator = new ItemEnqueuePositionCalculator(options);
+            DownloadStateProvider stubDownloadStateProvider = mock(DownloadStateProvider.class);
+            stub(stubDownloadStateProvider.isDownloadingFile(any(FeedMedia.class))).toReturn(false);
+            calculator.downloadStateProvider = stubDownloadStateProvider;
+
+            // Setup initial data
+            // A shallow copy, as the test code will manipulate the queue
+            List<FeedItem> queue = new ArrayList<>(queueInitial);
+
+
+            // Test body
+
+            Playable currentlyPlaying = getCurrentlyPlaying(idCurrentlyPlaying);
+
+            // User clicks download on feed item 101
+            FeedItem tFI101 = setAsDownloading(101, stubDownloadStateProvider);
+            doAddToQueueAndAssertResult(message + " (1st download)",
+                    calculator, tFI101, queue, currentlyPlaying,
+                    idsExpectedAfter101);
+
+            // Then user clicks download on feed item 102
+            FeedItem tFI102 = setAsDownloading(102, stubDownloadStateProvider);
+            doAddToQueueAndAssertResult(message + " (2nd download, it should preserve order of download)",
+                    calculator, tFI102, queue, currentlyPlaying,
+                    idsExpectedAfter102);
+
+            // TODO: simulate download failure cases.
```

Comment:
Please either remove the comment or implement the test

Reply:
test implemented - thanks for the reminder!